### PR TITLE
Fix RGBA image capturing error in vpFlyCaptureGrabber

### PR DESCRIPTION
**Pull request by xoox**
Two fix commits included:
1. Fix possible undefined behavior when memcpy
2. Fix error in RGBA capturing with FlyCapture SDK

---

Our cameras are on the way and expected arriving next month.
So these two commits are not fully tested. Could anybody have them reviewed and tested?

### CODE REVIEW

File: modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp [hunk 1]
```diff
@@ -1189,10 +1193,24 @@ void vpFlyCaptureGrabber::acquire(vpImage<vpRGBa> &I, FlyCapture2::TimeStamp &ti
...
+  // to an array of 32-bit RGBU values with each value a byte in the
+  // order of R, G, B and U and goes on.
+  for (unsigned int i = 0; i < height; ++i) {
+    for (unsigned int j = 0; j < width; ++j) {
+      unsigned char *pp = data + i * stride + j * Bps;
+      I[i * width + j]->R = pp[0];
+      I[i * width + j]->G = pp[1];
+      I[i * width + j]->B = pp[2];
+      I[i * width + j]->A = pp[3];
+    }
```
Comment (fspindle) [inline modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp]: Should be changes with:
````
  for (unsigned int i = 0; i < height; ++i) {
    for (unsigned int j = 0; j < width; ++j) {
      unsigned char *pp = data + i * stride + j * Bps;
      I[i][j].R = pp[0];
      I[i][j].G = pp[1];
      I[i][j].B = pp[2];
      I[i][j].A = pp[3];
    }
  }
````

**Comment by fspindle** [comment]
Acquiring `vpImage<vpRGBa>` produces a segfault. See code review for a fix.

Acquiring `vpImage<unsigned char>` produces the following error:
````
$ ./grabFlyCapture
Use device   : 0
Image size: 1280 960 1228800
sizeof(unsigned char): 1
debug 1
Error Trace:
Source: ImageConverter.cpp(154) Built: Mar 30 2017 01:41:49 - Conversion failed because the destination image would not accept the required dimensions.
+-> From: Image.cpp(448) Built: Mar 30 2017 01:41:36 - The buffer associated with this image is too small for the requested dimensions.
Catch an exception: Cannot convert image from camera with serial 15290004
````
I have a FLEA3-U3-13E4C-C USB3 camera configured to acquire 1280x960 images in MONO8 mode and I don't see what goes wrong in the proposed code

File: modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp [hunk 2]
```diff
@@ -1128,21 +1128,25 @@ void vpFlyCaptureGrabber::acquire(vpImage<unsigned char> &I, FlyCapture2::TimeSt
...
+  // char) * width`, which makes sure there is no paddings or holes
+  // between pixel data. And the convertedImage object is sharing the
+  // same data buffer with vpImage object `I`.
+  FlyCapture2::Image convertedImage(
+      height, width, sizeof(unsigned char) * width, I.bitmap,
+      sizeof(unsigned char) * I.getSize(), FlyCapture2::PIXEL_FORMAT_MONO8);
 
   // Convert the raw image
-  error = m_rawImage.Convert( FlyCapture2::PIXEL_FORMAT_MONO8, &convertedImage );
+  error = m_rawImage.Convert(&convertedImage);
```
Comment (xoox) [inline modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp]: Please have a try with line 1144 changed to
```c++
error = m_rawImage.Convert(FlyCapture2::PIXEL_FORMAT_MONO8, &convertedImage);
```

**Comment by xoox** [comment]
@fspindle Would have a try with line 1144 changed to
```c++
error = m_rawImage.Convert(FlyCapture2::PIXEL_FORMAT_MONO8, &convertedImage);
```

File: modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp (hunk 2 above)
Comment (fspindle) [inline modules/sensor/src/framegrabber/flycapture/vpFlyCaptureGrabber.cpp]: Perfect, it works.

**Comment by fspindle** [comment]
Once line 1144 changed we can merge this PR

**Comment by fspindle** [comment]
Could you please download, fill, scan and send us the copyright transfer form: <https://visp.inria.fr/contributing-code/>.
Thanks

**Comment by xoox** [comment]
Thanks. I've sent the copyright transfer agreement to the specified email.

**Pull request merged**